pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -183,39 +182,92 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.ServerResourcesInterf
 			return
 		}
 
-		// Something has changed, time to sync.
-		glog.V(2).Infof("syncing garbage collector with updated resources from discovery: %v", newResources)
-
 		// Ensure workers are paused to avoid processing events before informers
 		// have resynced.
 		gc.workerLock.Lock()
 		defer gc.workerLock.Unlock()
 
-		// Resetting the REST mapper will also invalidate the underlying discovery
-		// client. This is a leaky abstraction and assumes behavior about the REST
-		// mapper, but we'll deal with it for now.
-		gc.restMapper.Reset()
-
-		// Perform the monitor resync and wait for controllers to report cache sync.
-		//
-		// NOTE: It's possible that newResources will diverge from the resources
-		// discovered by restMapper during the call to Reset, since they are
-		// distinct discovery clients invalidated at different times. For example,
-		// newResources may contain resources not returned in the restMapper's
-		// discovery call if the resources appeared in-between the calls. In that
-		// case, the restMapper will fail to map some of newResources until the next
-		// sync period.
-		if err := gc.resyncMonitors(newResources); err != nil {
-			utilruntime.HandleError(fmt.Errorf("failed to sync resource monitors: %v", err))
-			return
-		}
-		// TODO: WaitForCacheSync can block forever during normal operation. Could
-		// pass a timeout channel, but we have to consider the implications of
-		// un-pausing the GC with a partially synced graph builder.
-		if !controller.WaitForCacheSync("garbage collector", stopCh, gc.dependencyGraphBuilder.IsSynced) {
-			utilruntime.HandleError(fmt.Errorf("timed out waiting for dependency graph builder sync during GC sync"))
-			return
-		}
+		// Once we get here, we should not unpause workers until we've successfully synced
+		attempt := 0
+		wait.PollImmediateInfinite(time.Second, func() (bool, error) {
+			attempt++
+
+			// On a reattempt, check if available resources have changed
+			if attempt > 1 {
+				newResources = GetDeletableResources(discoveryClient)
+				if len(newResources) == 0 {
+					glog.V(2).Infof("no resources reported by discovery, will retry (attempt %d)", attempt)
+					return false, nil
+				}
+			}
+
+			added := sets.NewString()
+			removed := sets.NewString()
+			for old := range oldResources {
+				if _, ok := newResources[old]; !ok {
+					removed.Insert(fmt.Sprintf("%+v", old))
+				}
+			}
+			for new := range newResources {
+				if _, ok := oldResources[new]; !ok {
+					added.Insert(fmt.Sprintf("%+v", new))
+				}
+			}
+
+			// Something has changed, time to sync.
+			glog.V(2).Infof("syncing garbage collector with updated resources from discovery (attempt %d). added: %v, removed: %v", attempt, added.List(), removed.List())
+
+			// Resetting the REST mapper will also invalidate the underlying discovery
```

Comment:
I don't think your restmapper is related to your discovery, is it?  Pretty sure this restmapper is shared for all controllers.

Reply:
this was an existing comment, and refers to the discovery client used by the restmapper:

https://github.com/kubernetes/kubernetes/blob/d02cf08e27f640f09ebd489e094176fd075f3463/staging/src/k8s.io/client-go/restmapper/discovery.go#L210-L220